### DISCUSSION TITLE
ValueError on Vectors.most_similar

### DISCUSSION THREAD

**Discussion by ZeeD** [explosion/spaCy#10465] [Help: Coding & Implementations]
Hi.
I'm trying to understand if it's possible to use spacy to find synonyms. I found that I could use `Vectors.most_similar`, but I'm having troubles with even a basic code.
```python
from spacy import load
import numpy

nlp = load('en_core_web_sm')
queries = numpy.asarray([numpy.random.uniform(-1, 1, (300,))])
most_similar = nlp.vocab.vectors.most_similar(queries, n=10)
```
This snippet is basically the example I found on the [most_similar documentation](https://spacy.io/api/vectors#most_similar), but if I run it, it ends with this error:

```Traceback (most recent call last):
  File "c:\Program Files\Python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\Program Files\Python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\vito.detullio\Desktop\workspace-srm-cognitive\_prove\spacynltk.py", line 6, in <module>
    most_similar = nlp.vocab.vectors.most_similar(queries, n=10)
  File "spacy\vectors.pyx", line 479, in spacy.vectors.Vectors.most_similar
ValueError: [E198] Unable to return 10 most similar vectors for the current vectors table, which contains 0 vectors.
```

can you help me understanding what I'm doing "wrong"?

**Comment by ZeeD** [comment]
Ok, found the "culprit" - I'm using [en_core_web_sm](https://spacy.io/models/en#en_core_web_sm) that has 

> 0 keys, 0 unique vectors (0 dimensions)

switching to a different model, such as [en_core_web_md](https://spacy.io/models/en#en_core_web_md) the error is gone

**Reply by polm** [reply]
Yes, the small models don't have word vectors.

**Comment by ZeeD** [comment]
Coming back to my "original" problem, I have written a simple script, to find word "synonims"

```python
from numpy import asarray
from spacy import load  # @UnresolvedImport

nlp = load('en_core_web_md')

words = ['cinema']
keyss, _, scoress = nlp.vocab.vectors.most_similar(asarray([nlp(word).vector
                                                            for word in words]),
                                                   n=10)
for word, keys, scores in zip(words, keyss, scoress):
    strings = [nlp.vocab.strings[key] for key in keys]
    print(f'word={word!r}')
    for string, score in zip(strings, scores):
        print(f'\tstring={string!r}, score={score!r}')
```

this snippet "works" - as in extract something.
my theoretical basis unfortunately is little, so I would like to know... what have I extracted? are word with "similar vectors" actually comparable to synonims?
Are the results (both "teoretically" and "pratically") comparable to the ones I could get, for example, from [wordnet synsets](https://www.nltk.org/howto/wordnet.html#synsets)?

**Reply by polm** [reply]
> Are the results (both "teoretically" and "pratically") comparable to the ones I could get, for example, from [wordnet synsets](https://www.nltk.org/howto/wordnet.html#synsets)?

No, they are not going to be synonyms in the traditional sense. The best you can hope for is that they are words that are used in similar contexts (= surrounding words). A simple example of how this differs from real synonyms is that antonyms in the same category, like "black" and "white", will usually be close in embedding space. 

There's a lot of research on this topic, you might search for "word vectors" and "synonyms" or "antonyms", or check the Jurafsky and Martin book.

https://web.stanford.edu/~jurafsky/slp3/
